BBM api addition docs

rburbidge opened the pull request:
First submit of BBM Platform docs.

rburbidge committed:
Adding BBM Platform docs.

jeffheifetz commented on api/blackberry_bbm.js:
```diff
@@ -0,0 +1,22 @@
+blackberry.bbm = {
+
+};
\ No newline at end of file
```
This file adds no new functionality and will create an empty page. I think it should be removed

rburbidge replied:
How can I link to the blackberry namespace if I remove this?

jeffheifetz replied:
I'm not sure what you mean by linking here, but its not like we enforce the docs to be valid Javascript

jeffheifetz commented:
I really like the tables you've put in to show the possible values, however this is something unlike any of our other docs. I'm not sure what people think whether this is something we should adopt and add this to other APIs, reject it, or accept inconsistency.

Another issue is that their style is so different from all of our own HTML

An example can be seen below

<img src="http://i.imgur.com/ogzYj.png" />

jeffheifetz commented on api/blackberry_bbm_platform_ui_menu.js:
```diff
@@ -0,0 +1,100 @@
+ * @namespace Provides the ability to add menu items within BBM. Menu items persist across device resets.
+ * @BB50+
+ */
+blackberry.bbm.platform.ui.menu = {
```
Don't you need some way to get the menu items you've already created? Like getMenuItems?

rburbidge replied:
We don't have it in Java right now. We plan to add it after RTM.

jeffheifetz commented:
Weren't Session and Channel supposed to have a common method for closing/ending for consistency sake?

jeffheifetz commented on api/blackberry_bbm_platform_io_Connection.js:
```diff
@@ -0,0 +1,288 @@
+     * <code>true</code> if the user is hosting on this connection; <code>false</code> otherwise.
+     * @BB50+
+     */
+    this.hosting = false;
```
needs an @type Boolean tag

jeffheifetz commented:
Since the creation of the propertyCB item we introduced the use of the @event tag and there is currently no use case for @propertyCB. Please convert all of them to @event tags.

For more information see Issue #122

jeffheifetz commented on api/blackberry_bbm_platform_io_IncomingJoinRequest.js:
```diff
@@ -0,0 +1,56 @@
+ * @extends blackberry.bbm.platform.io.JoinRequest
+ * @BB50+
+ */
+blackberry.bbm.platform.io.IncomingJoinRequest = function() {
```
Is there a reason there is no TOC entry?

rburbidge replied:
This is a small part of the hosted connections framework so I don't see too much merit.

I would have put a TOC entry for the two "add user" frameworks but it appears you can only have entries for whole objects. So I only have an entry for IO, Channel, and Session, hoping to lead them to the IO object.

jeffheifetz commented on api/blackberry_bbm_platform_io_OutgoingJoinRequest.js:
```diff
@@ -0,0 +1,40 @@
+ * @extends blackberry.bbm.platform.io.JoinRequest
+ * @BB50+
+ */
+blackberry.bbm.platform.io.OutgoingJoinRequest = function() {
```
Same question about TOC entry

jeffheifetz commented on api/blackberry_bbm_platform_io_Session.js:
```diff
@@ -0,0 +1,92 @@
+     * removed and {@link pendingUsers} will be set to <code>0</code>.
+     * <p>{@link onSessionEnded} will be called for all users currently in the connection.
+     * <p>The connection will become inactive for the current user, and should not be used afterwards.
+     * @throws IllegalStateException if the current user has already ended the connection.
```
Should be 

```
@throws {IllegalStateException} if the current user has already ended the connection.
```

jeffheifetz commented on api/blackberry_bbm_platform_io_Session.js:
```diff
@@ -0,0 +1,92 @@
+     * Removes the current user from the connection, but leaves it active for all other participants.
+     * <p>{@link onUserLeft} will be called for all other users currently in the connection.
+     * <p>The connection will become inactive for the current user, and should not be used afterwards.
+     * @throws IllegalStateException if the current user has already left the connection.
```
Should be 

```
@throws {IllegalStateException} if the current user has already left the connection.
```

jeffheifetz commented on api/blackberry_bbm_platform_self.js:
```diff
@@ -0,0 +1,92 @@
+         * or deny the change.
+         * @param {String} status The status: one of <code>"available"</code> or <code>"busy"</code>.
+         * @param {String} message The optional status message. If <code>null</code> then the default message of
+         * either "Available" or "Busy" will be used.
```
The two parameters disagree on the casing of the possible values, is this on purpose? Also no default value is provided on the first parameter, what happens with incorrect values?

rburbidge replied:
message is supposed to be optional. If they do not provide it then it defaults to "Available" or "Busy" based on the value of status.

ababut replied:
Let's tag the parameter as optional then

rburbidge replied:
Will do

jeffheifetz commented:
All of your namespace/classes need to have a beta tag addeed. It would look something like this

```
@beta For more information about the BBM API Beta ...
```

jeffheifetz commented on api/blackberry_bbm_platform_self_profilebox.js:
```diff
@@ -0,0 +1,192 @@
+     * 
+     * @BB50+
+     */
+    registerIcon : function(iconID, iconURI) {
```
What if the iconID already has an icon registered to it? I assume it overwrites, but this isn't documented

rburbidge replied:
It would overwrite it, but in practice they should never be overwriting an icon. I only allow it in case the developer wants to change around their icon-ID maps while in development.

ababut replied:
Same comment here that I'm making everywhere that I see IDs: the extension should be managing them internally. I cannot see a reason to ask the JS developer to manage IDs.

rburbidge replied:
Application developers can handle IDs.

ababut replied:
Java application developers handle IDs routinely :) JavaScript developers do not.

Typically, in JS you pass around an object that opaquely hides the implementation and has a few basic use-case oriented methods on it.

jeffheifetz commented on api/blackberry_bbm_platform_self_profilebox.js:
```diff
@@ -0,0 +1,192 @@
+     * @readOnly
+     * @BB50+
+     */
+    accessible : false,
```
Throughout the class you say "@throws {UserProfileBoxAccessException} If the device is in mass storage mode." but I wonder if in future there may be other reasons added. Might it not make more sense to reference this property which will list all of the possible reasons?

ababut replied:
Agreed...in the descriptions of the methods that throw the UserProfileBoxAccessException, we shouldn't mention that it's because the device is in MSM, rather that the Profile box is inaccessible. For the reasons why it might be inaccessible, they can refer to the description of this property and use it to test in code.

jeffheifetz commented on api/blackberry_bbm_platform_self_profilebox_ProfileBoxItem.js:
```diff
@@ -0,0 +1,54 @@
+ * @class A profile box item. Consists of text and possibly an icon.
+ * @BB50+
+ */
+blackberry.bbm.platform.self.profilebox.ProfileBoxItem = function() {
```
Why is there no TOC entry?

rburbidge replied:
There's a TOC for self.profilebox. I don't think that mentioning this object in TOC adds value.

jeffheifetz commented on api/blackberry_bbm_platform_ui.js:
```diff
@@ -0,0 +1,23 @@
+ */
+blackberry.bbm.platform.ui = {
+};
\ No newline at end of file
```
Another file to be removed

jeffheifetz commented on api/blackberry_bbm_platform_ui_menu.js:
```diff
@@ -0,0 +1,100 @@
+     * @param {Number} options.order Display order of this menu item; lower order items will be displayed first.
+     * @param {String} options.invitationMessage The invitation message displayed to the recipient.
+     * @throws {IllegalArgumentException} if <code>options.label</code> is not provided or <code>options.label.length > MAX_LENGTH_MENU_ITEM_LABEL</code>.
+     * @throws {IllegalArgumentException} if <code>options.invitationMessage</code> is not provided or <code>options.invitationMessage.length > MAX_LENGTH_MENU_ITEM_MESSAGE</code>.
```
If label and invitationMessage aren't optional, shouldn't they be required parameters outside of the JSON  and just leave optional parameters inside?

rburbidge replied:
All of the JSON parameters here are required though. I was recommended to use JSON for this method.

jeffheifetz commented on api/blackberry_bbm_platform_users.js:
```diff
@@ -0,0 +1,236 @@
+ * {@link blackberry.bbm.platform.self}.
+ * @BB50+
+ */
+blackberry.bbm.platform.users = {
```
This file seems to differentiate from the others and our conventions in terms of parameter order. The ordering should be:

```
1. Options Object
2. Non-optional arguments
3. Callbacks
4. Optional arguments
```

rburbidge replied:
Can you point out the methods with issues?

jeffheifetz commented on api/blackberry_bbm_platform_users_ContactInvitation.js:
```diff
@@ -0,0 +1,43 @@
+ * @constructor
+ * @BB50+
+ */
+blackberry.bbm.platform.users.ContactInvitation = function(PIN, name) {
```
Why is there no TOC entry?

rburbidge replied:
There are a few ways to invite users to your contact list. If possible I'd like to have them all under a single TOC entry. But they are split between objects so I'll need the ability to have TOC entries not based on objects.

jeffheifetz replied:
Well why are they split then. If you'd group the functionality under the TOC, it feels like it should be grouped under one namespace then.

rburbidge replied:
I was incorrect about that. They aren't split. The two methods are users.inviteToBBM and users.inviteToBBMFromConnections.

I tried to put a TOC entry on the two methods but it doesn't show up. Is there a way to put TOC entries for methods?

jeffheifetz commented on api/blackberry_bbm_platform_users_ContactInvitation.js:
```diff
@@ -0,0 +1,43 @@
+     * @readOnly
+     * @BB50+
+     */
+    this.name = "";
```
What does this represent? Is it the name I give to a contact on my list? Does it have any forced connection to the PIN?

rburbidge replied:
It's the display name in BBM which is independent of the PIN.

ababut commented:
I like the idea of presenting the possible values of in some format, but I'm not a fan of these Javadoc-style tables. 

As a starting point we can use a collapsible shaded section similar to that used for the whitelist features. It can be closed by default to avoid the visual clutter. 

What do you guys think?

ababut commented on api/blackberry_bbm_platform_ui_menu.js:
```diff
@@ -0,0 +1,100 @@
+     * inviter's side, {@link blackberry.bbm.platform.io.onConnectionCreated} will be invoked with
+     * the <code>event="BBMmenuinvite"</code> and <code>param=menuItemId</code>. If the application
+     * is not running then it will be launched. The invocation of
+     * {@link blackberry.bbm.platform.io.onConnectionCreated} will not occur until the callback is assigned.</p>
```
The statement is confusing/redundant - if it's a callback, it can't be invoked unless assigned. If the behaviour of that event is different with no callback assigned, it needs to be described in more detail. 

I recommend removing it from addInvitationMenuItem because it's confusing.

rburbidge replied:
Agreed. I will remove it.

ababut commented on api/blackberry_bbm_platform_ui_menu.js:
```diff
@@ -0,0 +1,100 @@
+    },
+
+    /**
+     * Returns <code>true</code> if the application has already added the menu item; <code>false</code> otherwise.
```
Is the menu item ID returned by addInvitationMenuItem?

rburbidge replied:
No the developer will set the menu item ID in the options parameter.

ababut commented on api/blackberry_bbm_platform_users.js:
```diff
@@ -0,0 +1,236 @@
+     * @description Shows Contact Picker dialog allowing the user to select contacts.
+     * @param {Object} options Object containing contact picker options.
+     * @param {String} [options.title] Title of the contact picker dialog.
+     * @param {String} [options.type] Type of users to include in the dialog. May be "contactswithapp".
```
What does "contactswithapp" do? Needs to get documented.

jeffheifetz replied:
I believe its a static readable property that gives you all of the contacts with the app installed. These can then be used for invites or other actions

ababut replied:
Gotcha, I see it now!

ababut commented on api/blackberry_bbm_platform_users.js:
```diff
@@ -0,0 +1,236 @@
+     * @param {Object} options Object containing contact picker options.
+     * @param {String} [options.title] Title of the contact picker dialog.
+     * @param {String} [options.type] Type of users to include in the dialog. May be "contactswithapp".
+     * @param {blackberry.bbm.platform.users.BBMPlatformUser[]} [options.users] Users shown in the
```
Does users have a default value/behaviour? If not, it should be reflected in the code example.

rburbidge replied:
For this API they should define either users, type, or both. Not sure how is best to convey that in the docs.

ababut replied:
What do they have to pass in for each of the categories?

rburbidge replied:
I'm not quite sure what you mean.

ababut replied:
You mentioned that they have to define users, type or both...so I'm guessing that the object literal you pass to pickUsers needs a type, users or both properties present.

When they are absent, what is their value assumed to be? We need to call out that at least one is required, and for the one not provided, what value it's assumed to have.

rburbidge replied:
The underlying java API has 3 overloaded methods: users, type, and users/type. So when one is missing it isn't really "defaulted" to anything, it just isn't used at all.

jeffheifetz commented:
I like the idea of collapsing but it would require template changes so they are always consistent.  I'm thinking something like @value {value} description.

However it does not address the issue of consistency unless we take on the extra work of adding this to the existing docs.

ababut commented on api/blackberry_bbm_platform.js:
```diff
@@ -0,0 +1,175 @@
+ * 
+ * <h1>Required BBM Version</h1>
+ * 
+ * BBM Social Platform APIs come with BBM6 and later.
```
You mention that the platform APIs come with BBM6 and later, but the supported platforms say 5.0+.

rburbidge replied:
BBM6 is the version of BBM, which is independent of BBOS version. I will add something like "BBM6 is supported on OS5, OS6, and OS7"

ababut replied:
Gotcha...in that case we should use branding-approved names. If BBM6 is it, ignore this comment, if not, pls find out what branding wants in there because we have to use that. For example, for instead of Playbook, they wanted our docs peppered with BlackBerry PlayBook :S

ababut commented on api/blackberry_bbm_platform.js:
```diff
@@ -0,0 +1,175 @@
+     * environment. If the application is in App World then the UUID will not be used. The same
+     * UUID should be used for future releases of the same application; otherwise communication
+     * between them will not be possible. The UUID must be a randomly generated 36-character UUID.
+     * Any UUID generator can be used.
```
Can we generate this UUID for them under the covers in Java? 

In general, we want to avoid asking JS developers to work with and manage IDs of any kind, unless absolutely necessary.

rburbidge replied:
Us generating it on their behalf could result in either inconsistent UUIDs or UUID collisions with other app developers. They need to know what their UUID is, it's super easy to generate one, and they only need to do it once.

Would it help if I put a link to an online UUID generator?

ababut replied:
No, we shouldn't be endorsing 3rd party tools directly. 

I'm not sure, but can the Java UID generator be seeded with a String value to always produce the same UUID?

rburbidge replied:
I'd really like to leave this as is since this is how it works on the Java side. Assigning a UUID shouldn't be too much of a struggle for app developers.

ababut commented on api/blackberry_bbm_platform.js:
```diff
@@ -0,0 +1,175 @@
+     * 
+     * // Register with the platform
+     * blackberry.bbm.platform.register({
+     *     uuid: "xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx" // Randomly generated UUID
```
Until we finalize this UUID, let's encourage them to use the WW method for generating one:

Check out Utils.generateUniqueId.

rburbidge replied:
The UUID needs to be the same on different runs of the same application. If it changes then all of the sudden contacts who have my app installed will no longer appear to have it installed because it is a different UUID. We can't trust this to Math.random().

ababut replied:
You're right, my bad!

ababut commented on api/blackberry_bbm_platform.js:
```diff
@@ -0,0 +1,175 @@
+     *     if(reason == "profilebox") {
+     *         var boxItem = param;
+     *         var achievementID = boxItem.cookie;
+     *         challenge(achievementID);
```
What does the challenge(acheievementID) method do? We should show it in the example to not confuse the copy-paste-pray masses :)

rburbidge replied:
I could replace that with a generic "// Take action based on profile box item" comment.

ababut replied:
Yes please...we want the sample code to not break anything when copy-pasted.

ababut commented on api/blackberry_bbm_platform_ui_menu.js:
```diff
@@ -0,0 +1,100 @@
+    MAX_LENGTH_MENU_ITEM_MESSAGE : 0,
+        
+    /**
+     * @description Adds a invitation menu item to the BBM contact list screen. If a menu item with
```
typo: "an invitation"

ababut commented on api/blackberry_bbm_platform_ui_menu.js:
```diff
@@ -0,0 +1,100 @@
+/*
+ * Copyright 2010 Research In Motion Limited.
```
The MenuItem API has a dangerous pitfall for WebWorks developers: they have to manage persistent IDs across application restarts and page switches. The trick is that BBM persists their menu entries, but unless they get a persistence mechanism in place, WW developers won't be able to track their menu items. Many simpler WW apps won't use any form of persistence as it adds a significant amount of complexity.

Also, IDs should be managed by the API for the reason above. I see a developer starting his app, incrementally using hasMenuItem to check integers from 0 to the first unused one and adding his item there. Then he has to persist that ID to use his menu item on the next page/app restart. 

This sort of low-level management should be handled by the extension. In the future, if BBM SDK's MenuItemManager starts managing it, you can't change the signature and pass the benefit downstream.

ababut replied:
I don't think that this is a showstopper for the Beta release, but I'm posting it so we have it on the backlog.

rburbidge replied:
There is no reason to persist the menu items IDs. The developer should not be adding menu items dynamically.

The must have a statically defined set of menu items (i.e. the same app on different devices should add the same menu items under the same IDs). This is so that the invitation can be correctly interpreted on the other side.

For example, all instances of a games app would define the following items regardless of device:

ID 1: checkers
ID 2: chess
ID 3: backgammon

If another app is assigning the wrong IDs then a "chess" invite might show up as a "checkers" invite on the other side.

A getMenuItems() API is planned but will not be available in this release. I think the API is fully functional as-is.

ababut replied:
What happens if I have another app, completely different from yours, and I add menu items to a contact:
1: Solitaire
2: Golf
3: Poker

Do I overwrite your items?

rburbidge replied:
Then those will show up under a different menu item... I should add a picture to the docs to show what the menu items look like.

ababut replied:
Yes please, or just update the description.

ababut commented on api/blackberry_bbm_platform_users_BBMPlatformUser.js:
```diff
@@ -0,0 +1,96 @@
+
+/**
+ * @featureID blackberry.bbm.platform
+ * @class A platform user.
```
We need to be a little more descriptive here :) 

If its a base class with informational fields that others inherit from, let's call that out. Can it be instantiated by itself?

rburbidge replied:
Good point. No you can't instantiate it yourself. I can add some information about where users are obtained from/how they are used.

ababut commented on api/blackberry_bbm_platform_users_BBMPlatformUser.js:
```diff
@@ -0,0 +1,96 @@
+     * @readOnly
+     * @BB50+
+     */
+    this.handle = 0;
```
Handle and PPID have the same description which is very confusing. I'm guessing that the handle is used to identify the device and the PPID the user?

If that's the case, we need to be specific. However, since the API doesn't support PlayBook, and I don't think that these values are used as parameters to any calls (I may have missed it though), do we need them?

rburbidge replied:
This is one of those concepts that is never easy to convey. I don't think that the docs for these can be improved any more. It already calls out the difference between them in bold.

ababut replied:
Hmm...how about something like:

handle - a unique, device-specific ID assigned to the user's account on each platform where it is used (BB, PB, etc)

ppid - a unique ID representing the user on all devices

rburbidge replied:
That's actually very good. I'll play with that and see what I come up with.

ababut commented on api/blackberry_bbm_platform_users_BBMPlatformUser.js:
```diff
@@ -0,0 +1,96 @@
+    this.personalMessage = "";
+
+    /**
+     * @description The status. One of <code>"available"</code> or <code>"busy"</code>.
```
I would make the description just: One of <code>"available"</code> or <code>"busy"</code>.

ababut commented on api/blackberry_bbm_platform_users_BBMPlatformUser.js:
```diff
@@ -0,0 +1,96 @@
+    this.status = "";
+
+    /**
+     * @description The status message.
```
I would make this description something like: The status message set by the user in BBM.

Basically to be clear that we're reading his status msg.

rburbidge replied:
Hmm, all of them are things that can be set in BBM or in third party apps. How about I add more information in the object level docs.

ababut replied:
Sounds good!

ababut commented on api/blackberry_bbm_platform_users_BBMPlatformUser.js:
```diff
@@ -0,0 +1,96 @@
+     * &lt;script type="text/javascript"&gt;
+     * // Show a contact's display picture
+     * var user = blackberry.bbm.platform.users.contactsWithApp[0];
+     * var displayPicImg = document.getElementById("displayPicture");
```
We should put a comment here saying that we're assuming that the user has an <img> tag with id='displayPicture'

ababut commented on api/blackberry_bbm_platform_users_BBMPlatformUser.js:
```diff
@@ -0,0 +1,96 @@
+     * // Show a contact's display picture
+     * var user = blackberry.bbm.platform.users.contactsWithApp[0];
+     * var displayPicImg = document.getElementById("displayPicture");
+     * displayPicImg.src = user.displayPicture;
```
Can both BB 5 and 6 take base-64 encoded data into the img tags? Some random blogger online was commenting that it only works in FF.

rburbidge replied:
They both work -- I've seen that the Contact.picture API works in both 5 and 6. However something is wrong with the way I base64 encode my images and it's currently not working in 5 with BBM API.

ababut replied:
Heh, the BB 5 browser's always the joker on the deck. If your base-64 encoded image works fine in BB6 and 7 but not 5, I'd suspect the browser. 

If that's the case, I suggest getting rid of the example altogether. Either it works for everyone, or they all figure it out on their own and we show a different example :)

rburbidge replied:
Yeah but as I said Contact.picture works in 5.0 and I'm convinced that I'll be able to get it working for my images too.

ababut commented on api/blackberry_bbm_platform_users_BBMPlatformUser.js:
```diff
@@ -0,0 +1,96 @@
+     * displayPicImg.src = user.displayPicture;
+     * &lt;/script&gt;
+     * 
+     * @example
```
Duplicate example.

rburbidge replied:
They aren't duplicates -- one shows display picture of current user, the other shows for a contact with the app. Which would you like me to keep?

ababut replied:
Gotcha, I just saw that...I say we keep the self picture, the code example only references items on that docs page. The other one could be confusing because they might not know where contactsWithApp[0] came from.

ababut commented on api/blackberry_bbm_platform_self.js:
```diff
@@ -0,0 +1,92 @@
+         * </p>
+         * @param {String} displayPictureURI The fully qualified URI.
+         * @callback {Function} onComplete Invoked when the user has dismissed the dialog.
+         * @callback {Boolean} onComplete.result <code>true</code> if the user allowed the change;
```
Let's call the callback param "accepted" or "hasAccepted"...the example code would read more naturally.

rburbidge replied:
Good suggestion

ababut commented on api/blackberry_bbm_platform_self.js:
```diff
@@ -0,0 +1,92 @@
+        /**
+         * @description Sets the user's personal message. A dialog will be presented to the user to
+         * allow or deny the change.
+         * @param {String} personalMessage The personal message. The maximum length is 64 characters;
```
I suggest not mentioning the upper limit. It gives you freedom to change it in the future. This will encourage developers to add logic to check that the message isn't longer than 64 characters.

ababut replied:
Also, if not too late, let's not encourage them to pass in null. Would empty string ("") also clear it?

rburbidge replied:
If the text randomly gets truncated the devs will be angry because they weren't told why.

Why don't we want developers using null?

ababut replied:
It leaves the door open to potential bugs coming from NullPointerExceptions. Empty Strings are safe.

It's up to you for the message, but telling people the maximum length will get them coding against it.

ababut commented on api/blackberry_bbm_platform_self_profilebox.js:
```diff
@@ -0,0 +1,192 @@
+ * <h3>Icon IDs</h3>
+ * <ul>
+ * <li>Each icon must be given a unique non-negative ID.
+ * <li>Each icon ID must map to the same icon on all devices. Images generated at runtime such as
```
Is there a size constraint on images generated at runtime?

rburbidge replied:
No. Any images they provide will be scaled anyway.

ababut replied:
Then can we use an image at runtime?

rburbidge replied:
Yes they can retrieve the icon using self.profilebox.getIcon()

ababut commented on api/blackberry_bbm_platform_self_profilebox.js:
```diff
@@ -0,0 +1,192 @@
+    items : null,
+    
+    /**
+     * Registers a new icon from a URI.
```
Is it possible to not impose the registration of icons at the beginning constraint? If it's for performance reasons, we can list it as a best practice, but registering an icon before adding it implies ordering of calls. 

What happens if they add an item and reference an ID that doesn't exist?

rburbidge replied:
The reason for registering before is so that I don't have to download icons if viewing the same app on another user's profile. It isn't absolutely necessary, but it's a better user experience.

If you reference the ID on a not-registered icon then it just shows up as no icon in the profile.

ababut commented on api/blackberry_bbm_platform_self_profilebox.js:
```diff
@@ -0,0 +1,192 @@
+     * @example
+     * &lt;script type="text/javascript"&gt;
+     * // Add an item with icon, text, and a cookie
+     * var options = {text:"Planted an apple orchard on 10 acres!", iconId:ICON_INFO.apple.id, cookie:"10acres"};
```
ICON_INFO.apple.id is coming from the example for registerIcon. Since there several steps involved with updating the profileBox, we should consolidate all per-method examples into a single coherent example for the class. 

It will serve to bring together the process' various steps.

rburbidge replied:
I can do that.

rburbidge committed:
Updating docs with feedback

ababut commented on api/blackberry_bbm_platform_users.js:
```diff
@@ -0,0 +1,236 @@
+    contactsWithApp : [],
+
+    /**
+     * @description Shows Contact Picker dialog allowing the user to select contacts.
```
This description needs to be updated - what are the picked users used for?

rburbidge replied:
They can be used for anything. In my sample app I use pickUsers so that the user can select others to remove from a connection.

ababut replied:
Gotcha, I understand now.

rburbidge committed:
Changed UserProfileBoxAccessException to generic description. Removed bbm and ui namespace files.

ababut commented on api/blackberry_bbm_platform_users.js:
```diff
@@ -0,0 +1,236 @@
+     * // Start chat without Contact Picker dialog
+     * blackberry.bbm.platform.users.startBBMChat(function() {
+     *     // Continue with application...
+     * }, "I won by 500 points!", [player1, player2, player3]);
```
The players array will break this example.

rburbidge replied:
Then the only example I could really show is the top one.

Is it necessary that all examples must work if copy/pasted? Sometimes this will never be possible without straying from the point of the API. (case in point)

This may sound caustic, but I'm not sure if this audience of copy/paste developers is the kind we want to attract to our platform.

ababut replied:
The issue is that an error in JavaScript immediately stops the script engine. To the user it looks like our code crashed their app. We had a number of situations when developers used our non-self-contained snippets, their app crashed and we received negative feedback. 

I totally understand the BBM API's scope, and I know a lot of pieces tie together, but the onus is still on us to provide non-crashing examples. This means combining several smaller examples into a big, class-level one if that makes sense. Or we can do what you did earlier and put comments replacing code. Whatever makes sense in a particular situation.

rburbidge replied:
I will remove the second example.

ababut commented on api/blackberry_bbm_platform_users.js:
```diff
@@ -0,0 +1,236 @@
+     *         var userHandle = user.handle;
+     *         if (userHandle == blackberry.platform.self.handle) {
+     *             displayPicId = "myPicture";
+     *         } else if (userHandle == p2Handle) {
```
pXHandle will break this example.

rburbidge committed:
Moved io.sendFile() and io.inviteToDownload() to the users namespace.

rburbidge commented:
About the tables: I can always add the class="confluenceTh" and class="confluenceTd" attributes to each cell in the tables. Although it still leaves to docs open for inconsistency if you ever choose to change the style of the docs.

Alternatively I could use lists with the item being described in bold.

rburbidge committed:
Fixed broken examples:
Removed broken example for users.startBBMChat()
Fixed broken example for users.onUpdate()

rburbidge committed:
Added two @learn tags in platform namespace.
Added @beta tags in every namespace/object.

rburbidge committed:
Converted @propertyCB to @event for platform namespace. Made callbacks lower-case.

rburbidge committed:
Updated users.onUpdate with @event tag, lowercase.

rburbidge committed:
Fixed code examples for new lowercase onaccesschanged.

rburbidge committed:
Updated Connection and Session callbacks with @event tag, lowercase.

rburbidge committed:
Changed io namespace callbacks to lowercase. Split io.onConnectionCreated into two callbacks: io.onconnectionaccepted and io.onbbmmenuinvite.

rburbidge committed:
Updated profile box API:
Removed the icon registration APIs + docs. This is now handled automatically in profilebox.addItem()
profilebox.registerIcon()
profilebox.getIcon()
profilebox.hasIcon()
Icons can now be retrieved from ProfileBoxItem.icon.

rburbidge committed:
Fixed a few event links in the platform and profilebox namespaces.

rburbidge committed:
Added depraction notice to all menu APIs.

rburbidge committed:
Added missing status/reason strings:
- platform.onaccesschanged() "nonuiapp" status
- users.sendFile() "fileempty" reason

rburbidge committed:
Fixed blackberry.bbm.platform.users.pickUsers() code sample. Previously read "blackberry.bbm.platform.contacts.pickUsers"

rburbidge committed:
Added more top level docs for BBMPlatformUser and blackberry.bbm.platform.users. Added code samples for users.sendFile() and users.inviteToDownload().

rburbidge committed:
Added requestUserPermission API.

rburbidge committed:
Corrected examples of how to send/receive JSON data.

rburbidge committed:
Updating images to match new lower-case event callbacks.

rburbidge committed:
Changed table of contents to a single BBM section with ordered links.

rburbidge committed:
Removed empty blackberry.bbm namespace file.
Removed blackberry.bbm.platform.ui and blackberry.bbm.platform.ui.menu namespaces as these are not being included at the moment.

rburbidge committed:
Removed empty blackberry.bbm namespace file.

rburbidge committed:
Merge branch 'epsilon' of https://github.com/rburbidge/WebWorks-API-Docs into epsilon

rburbidge committed:
Reworked BBM TOC.

rburbidge committed:
Removed io.onbbmmenuinvite.

rburbidge committed:
Fixed broken event link in Session.end() docs.

rburbidge committed:
Added characters to constants in Connection object.

rburbidge committed:
Changed getting started guide to the main link BBM WebWorks GS guide. Added other guide as supplementary.

rburbidge committed:
Improved the docs for platform.onappinvoked. Added notes for platform.onaccesschanged and io.onconnectionaccepted that these callbacks need to be assigned before the call to register.

rburbidge committed:
Changed blackberry.bbm.platform.users.inviteToBBM() to use anonymous object instead of ContactInvitation object. Removed blackberry.bbm.platform.users.ContactInvitation object alltogether.

rburbidge committed:
Refactored hosting methods/events to io namespace:
- Connection.enableHosting() -> io.host(...)
- Connection.disableHosting() -> io.host()
- Connection.incomingJoinRequests -> io.hostRequests
- io.requestToJoin() -> io.joinHost()
- io.outgoingJoinRequests -> io.joinHostRequests

Refactored events so that callbacks are supplied directly to host/joinHost:
- Connection.onjoinrequestreceived and Connection.onjoinrequestcanceled removed, now supplied directly to io.host()
- io.onjoinrequestaccepted and io.onjoinrequestdeclined removed, now supplied directly to io.joinHost()

- Connection.hosting removed
- io.hostedConnection added

rburbidge committed:
Added blurb about avoiding the Contact Picker dialog. Corrected some instances of contact picker->Contact Picker. Added headings in namespace level docs.

rburbidge committed:
Changed Channel.close() to leave(). This matches Session.leave() and performs the same function.

rburbidge committed:
Added missing exceptions for Connection and Session. Added blurb about Unreachable users in Connection.send() and Session.broadcast().

rburbidge committed:
Changed io.host() so that if a connection is already being hosted, then calling host() on a different connection will stop hosting on the orignial and begin hosting on the new connection.

rburbidge committed:
Replaced constant string tables with lists. Made minor updates to io namespace. Added missing reason parameter for the onJoinRequestCanceled callback in io.host().

rburbidge committed:
Changed the name of the callback parameters to io.host(). Added note to OutgoingJoinRequest.cancel() to specify which callback will be invoked on the host's side.

rburbidge committed:
Added more detail on stopping hosting in io.host() docs.

rburbidge committed:
Corrected names of io.host() callbacks.

rburbidge committed:
Updated hosting diagram.
